### ISSUE TITLE
Cannot connect

### ISSUE THREAD

**Issue by mreese84** [tcampb/roadlight#2]
I have the app running locally, but with TrainerRoad running and my Hue credentials/IP address populated, the app won't connect to either TrainerRoad or my lights. Screenshots included. Please let me know if this is simply user error or if there's an issue in the code. Notably, when I look at the console I see three errors pop up. Thank you! Can't wait to get this thing running.
<img width="1440" alt="Screen Shot 2023-08-14 at 5 18 40 PM" src="https://github.com/tcampb/roadlight/assets/7680119/cc509985-2ca9-45da-9e57-3e01b65037d7">
<img width="1440" alt="Screen Shot 2023-08-14 at 5 26 03 PM" src="https://github.com/tcampb/roadlight/assets/7680119/ee15cfb1-b9d7-4f93-a65e-33f5fe7c507e">

**Comment by tcampb** [comment]
Hey @mreese84,

1) For the trainerroad error, could you confirm that you had trainerroad open and were logged in prior to opening roadlight?

2) For the philips hue bridge error, did you create a user in the bridge as shown in this video ([https://www.youtube.com/watch?v=CijfE7AFNEM](https://www.youtube.com/watch?v=CijfE7AFNEM))? The philips hue instructions start at 57 seconds.

I had forgotten to make the video above public, so apologies for the lack of setup instructions :sweat_smile: 


Also, not sure if you already use the Hue Sync app, but based on the trainerroad comment [here](https://www.trainerroad.com/forum/t/getting-lights-in-room-to-match-colour-of-tr-effort/8271/16) it sounds like this can also be setup within the app.

**Comment by mreese84** [comment]
@tcampb,
1. Yes... I open TR, let the app load completely, and confirm I am logged in (note, I am already connected to my Kickr as a power meter. Not sure if that matters or not. I tried doing it with my Kickr unplugged and still got an error.) I then open my terminal, run `npm start` and the Electron app opens. I get an error immediately saying it can't load Hue lights and another error saying it can't connect to TR.
2. As for the hue error... definitely did not follow those instructions. I was using my normal hue username and IP address from the philips hue app. I'll follow those steps now.
Thanks for the quick reply!

**Comment by mreese84** [comment]
@tcampb, okay... I'm connected to the bridge. Those instructions were great. Just need to solve the TR connection issue now.

**Comment by mreese84** [comment]
A couple more clues for you...

1. I'm on macOS 11.7.4
2. I'm using npm 9.8.0
3. TrainerRoad version is 2023.31.1.265
4. When I open Dev Tools, I see this error in the console...
<img width="1552" alt="Screen Shot 2023-08-14 at 8 45 24 PM" src="https://github.com/tcampb/roadlight/assets/7680119/c1b5d9de-1c89-4497-8f46-d4bc6590f37f">

Thanks for taking a look!

**Comment by tcampb** [comment]
Thanks for all of the additional information @mreese84, this is very helpful! It looks like a recent update to the Trainerroad client changed the name of the process that outputs power information. Just wanted to give you a heads up that I am looking into this now and will hopefully have an update soon.

**Comment by mreese84** [comment]
@tcampb fantastic. Thank you!

**Comment by tcampb** [comment]
Hey @mreese84, it looks like the `EmbeddedDeviceService` that roadlight connects to is no longer started automatically when the Trainerroad application is run. It looks like they now connect directly to the bluetooth / ant device file descriptor from what I can tell.

That said, it looks like the `EmbeddedDeviceService` is still packaged in the Trainerroad application. I was able to connect to it by manually starting the `EmbeddedDeviceService`:

```bash
cd /Applications/TrainerRoad.app/Contents/Resources/EmbeddedDeviceService/Mac/TrainerRoad.Embedded.Mac.app/Contents/MacOS/TrainerRoad.Embedded.Mac

./TrainerRoad.Embedded.Mac --version 1 --channel channel
``` 

I'm not sure what exactly the `version` and `channel` arguments do, but they are required so I just passed in a random value.

**Comment by mreese84** [comment]
Hmm, weird. I was able to start `EmbeddedDeviceService` and get rid of the error. So that's great. However, Roadlight isn't getting power numbers from TR. It says that it is connected, but nothing is synced. I'm not seeing any warnings or errors anywhere in the console or terminal either. Any ideas?

**Comment by tcampb** [comment]
Ah, that makes sense. Since it's not being started by the Trainerroad application it is not receiving the power info.

I'm thinking at this point it probably makes sense to drop the EmbeddedDeviceService dependency and instead use the [web bluetooth api](https://developer.mozilla.org/en-US/docs/Web/API/Web_Bluetooth_API) to connect to the power meter. 

It's been a while since I've done any electron development so I'll have to do some research :sweat_smile: 

Do you know if your Kickr can be paired with multiple devices at once, for example Trainerroad and Roadlight so that they both can receive the power data?

**Comment by mreese84** [comment]
Ahh, bummer. That's unfortunate. As for your question, I just tested it and yes I am reading power from my Kickr on both TR and a Garmin watch. So I take that as a good sign the Kickr could connect to both TR and Roadlight.

**Comment by mreese84** [comment]
Additionally, if we're talking about connecting directy to the trainer/power meter, I think this would be better than connecting through an app like TR because if you connect directly to the Kickr, that opens up usability to anyone using a bluetooth-capable power meter, not just TrainerRoad users. This is something the cycling/hue-user world have wanted for a while. You'd be a legend if you made this happen!

**Comment by tcampb** [comment]
Thanks @mreese84, this is really helpful information! I was able to create a quick proof of concept to connect over Bluetooth instead of using Trainerroad. 

I'll try and push up those changes for testing tomorrow. I had originally stopped working on this since I thought an app already existed, so this is really helpful info!

**Comment by mreese84** [comment]
@tcampb fantastic! Looking forward to it. The Hue Sync thing is an _okay_ stopgap, but not what people are really after. The ability to customize colors like you have (even though it would require altering the code) is great. With Sync you're stuck with whatever color whatever app has decided to use. I love the ability to set different values for different power zones (and eventually heart rate ;-) ). Setting colors from within the UI would be killer, but beggers can't be choosers. I'm happy to poke around in the code to update those myself.

**Comment by tcampb** [comment]
@mreese84 - I've just pushed up the code changes for the new version here if you want to try it out:

https://github.com/tcampb/roadlight-web

If it works I'll archive this repository and link new users to the roadlight-web version. This version runs a web application instead of a desktop app since it no longer depends on Trainerroad.

**Comment by mreese84** [comment]
Awesome. I just pulled it down, but the app doesn't start up when I run `npm start`. Not sure how to get the UI open.

**Comment by tcampb** [comment]
Hmm, I just cloned the repo and ran through the steps on a separate device and it worked.

Can you try manually running the npm start command:

```bash
node app.js
```

This should start a web server listening on port 3030

http://localhost:3030

**Comment by mreese84** [comment]
Ahh, I was expecting an Electron app window to fire. I gotchu. Testing now!

**Comment by mreese84** [comment]
Ok, app loads fine. I'm able to connect to my Hue lights and pair with my power meter. However, when I pedal the app isn't actually getting power numbers. I've tried both with TrainerRoad running and not running. Any thoughts?

**Comment by tcampb** [comment]
When you click "Connect" you should get a prompt to select a bluetooth power meter to connect to:

![image](https://github.com/tcampb/roadlight/assets/33756113/89c0d86a-b995-4a4d-82a1-d679dca09863)

**Comment by mreese84** [comment]
Yeah. I've connected and it says that I'm paired. It's just that no power readings come across when I pedal.

**Comment by tcampb** [comment]
Are you seeing any output in terminal or in the [browser console](https://support.monday.com/hc/en-us/articles/360002197259-How-to-open-the-developer-console#:~:text=you're%20using.-,Chrome,to%20open%20the%20developer%20console.)? Also, I sent a LinkedIn message if you'd like to coordinate setting up a video call to debug.
